tools/normalize.c

```diff
@@ -0,0 +1,177 @@
+ * limitations under the License.
+ */
+
+#include <errno.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <inttypes.h>
+#include <string.h>
+
+#include "../src/msg.h"
+
+#include "msg.h"
+#include "sig.h"
+#include "fifo.h"
+
+// DEFAULT_AVERAGE_WIDTH, the number of milliseconds to average over
+long DEFAULT_AVERAGE_WIDTH = 1000;
+
+struct channel_data {
+	bl_msg_sample_data_t msg; // Used as a buffer and for outputting the processed samples
+	struct fifo samples;      // Store samples until we've finished using them for processing
+	double rolling_average;   // The rolling average as of the most recent sample
+};
+
+void add_normalized_sample(struct channel_data *channel, uint32_t sample) {
+	// If adding it to the message is impossible because it's full, write the message.
+	// sample may be 16-bit or 32-bit full of numbers, depending on data format.
+	if (channel->msg.type == BL_MSG_SAMPLE_DATA16) {
+		channel->msg.data16[channel->msg.count] = (uint16_t) sample;
+		channel->msg.count++;
+	} else {
+		channel->msg.data32[channel->msg.count] = sample;
+		channel->msg.count++;
+	}
+	if ((channel->msg.type == BL_MSG_SAMPLE_DATA16 && channel->msg.count == MSG_SAMPLE_DATA16_MAX)
+	    || (channel->msg.type == BL_MSG_SAMPLE_DATA32 && channel->msg.count == MSG_SAMPLE_DATA32_MAX)) {
+			// Message is full, send and start refilling
+			bl_msg_yaml_print(stdout, (union bl_msg_data *) &channel->msg);
+			channel->msg.count = 0;
+	}
+}
+
+int add_sample(long average_width, struct channel_data *channel, uint32_t sample)
+{
+	// Add a sample to the fifo
+	if (!fifo_write(&channel->samples, sample)) {
+		fprintf(stderr, "FIFO overflow\n");
+		return -ERANGE;
+	}
+	// Add to the rolling average
+	channel->rolling_average += (float) sample / (float) average_width;
+	if (channel->samples.used >= average_width) {
+		uint32_t new_sample, old_sample;
+		double normalized_sample, sample_ratio;
+
+		// Calculate the normalized sample
+	   	if (!fifo_peek_back(&channel->samples, average_width / 2, &new_sample)) {
+			fprintf(stderr, "FIFO underflow\n");
+			return -ERANGE;
+		}
+		sample_ratio = (float) new_sample / channel->rolling_average;
+		if (channel->msg.type == BL_MSG_SAMPLE_DATA16) {
+			normalized_sample = sample_ratio * UINT16_MAX / 2;
+		} else {
+			normalized_sample = sample_ratio * UINT32_MAX / 2;
+		}
+
+		// Write the normalized sample
+		add_normalized_sample(channel, (uint32_t) normalized_sample);
+
+		// Subtract from the rolling average
+		if (!fifo_read(&channel->samples, &old_sample)) {
+			fprintf(stderr, "Tried to read from empty FIFO\n");
+			return -ENOENT;
+		}
+		channel->rolling_average -= (float) old_sample / (float) average_width;
+	}
+	return 0;
+	
+}
+
+void init_channel(struct channel_data *channel, union bl_msg_data *msg) {
+	memset(channel, 0, sizeof(*channel));
+	channel->msg.type = msg->channel_conf.sample32 ? BL_MSG_SAMPLE_DATA32 : BL_MSG_SAMPLE_DATA16;
+	channel->msg.channel = msg->sample_data.channel;
```

Comment:
Since it's a union, the message is either `channel_conf` or `sample_data`.  We can't read both in here.

Since we're calling this from where we read the channel conf message, I'd be tempted to change the prototype to

```c
void init_channel(struct channel_data *channel, const bl_msg_channel_conf_t *channel_conf);
```

We can get the channel index and the 32-bitedness from there, and the tightened up type helps reduce the chances of errors.